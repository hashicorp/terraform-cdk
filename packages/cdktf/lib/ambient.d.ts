// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
declare module 'archiver';