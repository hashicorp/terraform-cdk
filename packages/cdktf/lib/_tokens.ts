// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, DefaultTokenResolver, StringConcat } from "./tokens";
import { Construct } from "constructs";

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

// eslint-disable-next-line jsdoc/require-jsdoc
export function resolve<T>(scope: Construct, obj: T, preparing = false): T {
  return Tokenization.resolve(obj, {
    scope,
    preparing,
    resolver: TOKEN_RESOLVER,
  });
}
