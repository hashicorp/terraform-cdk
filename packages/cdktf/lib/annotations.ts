// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
// Originally from aws-cdk v2 because with constructs v10 addMessage was removed
// in favor of Annotations in the AWS CDK
// https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/annotations.ts
import { IConstruct } from "constructs";

export enum AnnotationMetadataEntryType {
  INFO = "@cdktf/info",
  WARN = "@cdktf/warn",
  ERROR = "@cdktf/error",
}

export const DISABLE_STACK_TRACE_IN_METADATA =
  "@cdktf/disable_stack_trace_in_metadata";

/**
 * Includes API for attaching annotations such as warning messages to constructs.
 */
export class Annotations {
  /**
   * Returns the annotations API for a construct scope.
   * @param scope The scope
   */
  public static of(scope: IConstruct) {
    return new Annotations(scope);
  }

  private readonly stackTraces: boolean;

  private constructor(private readonly scope: IConstruct) {
    const disableTrace =
      scope.node.tryGetContext(DISABLE_STACK_TRACE_IN_METADATA) || false;

    this.stackTraces = !disableTrace;
  }

  /**
   * Adds a warning metadata entry to this construct.
   *
   * The CLI will display the warning when an app is synthesized.
   * In a future release the CLI might introduce a --strict flag which
   * will then fail the synthesis if it encounters a warning.
   *
   * @param message The warning message.
   */
  public addWarning(message: string) {
    this.addMessage(AnnotationMetadataEntryType.WARN, message);
  }

  /**
   * Adds an info metadata entry to this construct.
   *
   * The CLI will display the info message when apps are synthesized.
   *
   * @param message The info message.
   */
  public addInfo(message: string): void {
    this.addMessage(AnnotationMetadataEntryType.INFO, message);
  }

  /**
   * Adds an { "error": <message> } metadata entry to this construct.
   * The toolkit will fail synthesis when errors are reported.
   * @param message The error message.
   */
  public addError(message: string) {
    this.addMessage(AnnotationMetadataEntryType.ERROR, message);
  }

  /**
   * Adds a message metadata entry to the construct node, to be displayed by the CDK CLI.
   * @param level The message level
   * @param message The message itself
   */
  private addMessage(level: AnnotationMetadataEntryType, message: string) {
    this.scope.node.addMetadata(level, message, {
      stackTrace: this.stackTraces,
    });
  }
}
