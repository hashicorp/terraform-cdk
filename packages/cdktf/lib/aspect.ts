// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
//  Originally from aws-cdk v2 because with constructs v10 Aspects where moved to the AWS CDK
//  https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/aspect.ts
import { IConstruct } from "constructs";

const ASPECTS_SYMBOL = Symbol("cdktf-aspects");

/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * All aspects can visit an IConstruct
   */
  visit(node: IConstruct): void;
}

/**
 * Aspects can be applied to CDK tree scopes and can operate on the tree before
 * synthesis.
 */
export class Aspects {
  /**
   * Returns the `Aspects` object associated with a construct scope.
   * @param scope The scope for which these aspects will apply.
   */
  public static of(scope: IConstruct): Aspects {
    let aspects = (scope as any)[ASPECTS_SYMBOL];
    if (!aspects) {
      aspects = new Aspects();

      Object.defineProperty(scope, ASPECTS_SYMBOL, {
        value: aspects,
        configurable: false,
        enumerable: false,
      });
    }
    return aspects;
  }

  private readonly _aspects: IAspect[];

  private constructor() {
    this._aspects = [];
  }

  /**
   * Adds an aspect to apply this scope before synthesis.
   * @param aspect The aspect to add.
   */
  public add(aspect: IAspect) {
    this._aspects.push(aspect);
  }

  /**
   * The list of aspects which were directly applied on this scope.
   */
  public get all(): IAspect[] {
    return [...this._aspects];
  }
}
