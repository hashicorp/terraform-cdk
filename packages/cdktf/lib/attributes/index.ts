export * from "./terraform-any-attribute"
export * from "./terraform-any-list-attribute"
export * from "./terraform-any-map-attribute"
export * from "./terraform-any-set-attribute"
export * from "./terraform-attribute"
export * from "./terraform-boolean-attribute"
export * from "./terraform-boolean-list-attribute"
export * from "./terraform-boolean-map-attribute"
export * from "./terraform-boolean-set-attribute"
export * from "./terraform-dynamic-attribute"
export * from "./terraform-list-attribute"
export * from "./terraform-map-attribute"
export * from "./terraform-number-attribute"
export * from "./terraform-number-list-attribute"
export * from "./terraform-number-map-attribute"
export * from "./terraform-number-set-attribute"
export * from "./terraform-object-attribute"
export * from "./terraform-set-attribute"
export * from "./terraform-string-attribute"
export * from "./terraform-string-list-attribute"
export * from "./terraform-string-map-attribute"
export * from "./terraform-string-set-attribute"
