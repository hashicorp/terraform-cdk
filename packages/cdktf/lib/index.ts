// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { validateEnvironment } from "./validateEnvironment";
validateEnvironment();

export * from "./tokens";
export * from "./terraform-stack";
export * from "./terraform-element";
export * from "./terraform-module";
export * from "./terraform-provider";
export * from "./terraform-resource";
export * from "./terraform-data-source";
export * from "./terraform-output";
export * from "./complex-computed-list";
export * from "./resource";
export * from "./testing";
export * as testingMatchers from "./testing/matchers";
export * from "./app";
export * from "./terraform-backend";
export * from "./backends";
export * from "./terraform-remote-state";
export * from "./terraform-local";
export * from "./terraform-variable";
export * from "./terraform-hcl-module";
export * from "./runtime";
export * from "./terraform-dependable";
export * from "./manifest";
export * from "./terraform-asset";
export * from "./synthesize";
export * from "./annotations";
export * from "./aspect";
export * from "./terraform-functions";
export * from "./terraform-operators";
export * from "./tfExpression";
export * from "./terraform-addressable";
export * from "./terraform-iterator";
export * from "./terraform-provisioner";
export * from "./terraform-conditions";
export * from "./terraform-count";

// required for JSII because Fn extends from it
export * from "./functions/terraform-functions.generated";
