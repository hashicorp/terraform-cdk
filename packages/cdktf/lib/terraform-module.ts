import { Construct } from "constructs";
import { TerraformElement } from "./terraform-element";

export interface TerraformModuleOptions {
  readonly source: string;
  readonly version: string;
}

export abstract class TerraformModule extends TerraformElement {

  public readonly source: string;
  public readonly version: string;

  constructor(scope: Construct, id: string, options: TerraformModuleOptions) {
    super(scope, id);

    this.source = options.source;
    this.version = options.version;
  }

  // jsii can't handle abstract classes?
  protected synthesizeAttributes(): { [name: string]: any } {
    return {}
  }

  public interpolationForOutput(moduleOutput: string) {
    return `\${module.${this.friendlyUniqueId}.${moduleOutput}}` as any;
  }

  public toTerraform(): any {
    const attributes = {
      ...this.synthesizeAttributes(),
      source: this.source,
      version: this.version,
      "//": this.nodeMetadata
    }

    return {
      module: {
        [this.friendlyUniqueId]: attributes
      }
    }
  }
}