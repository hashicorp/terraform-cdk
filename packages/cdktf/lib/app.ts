// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Construct, IConstruct } from "constructs";
import * as fs from "fs";
import { version } from "../package.json";
import { DISABLE_STACK_TRACE_IN_METADATA } from "./annotations";
import { Manifest } from "./manifest";
import { ISynthesisSession } from "./synthesize";
import { TerraformStack } from "./terraform-stack";

const APP_SYMBOL = Symbol.for("cdktf/App");
export const CONTEXT_ENV = "CDKTF_CONTEXT_JSON";
export interface AppConfig {
  /**
   * The directory to output Terraform resources.
   *
   * @default - CDKTF_OUTDIR if defined, otherwise "cdktf.out"
   */
  readonly outdir?: string;
  readonly stackTraces?: boolean;

  /**
   * Additional context values for the application.
   *
   * Context set by the CLI or the `context` key in `cdktf.json` has precedence.
   *
   * Context can be read from any construct using `node.getContext(key)`.
   *
   * @default - no additional context
   */
  readonly context?: { [key: string]: any };

  /**
   * Whether to skip all validations during synthesis of the app
   *
   * @default - false
   */
  readonly skipValidation?: boolean;

  /**
   * Whether to skip backend validation during synthesis of the app
   *
   * @default - false
   */
  readonly skipBackendValidation?: boolean;
}

/**
 * Represents a cdktf application.
 */
export class App extends Construct {
  /**
   * The output directory into which resources will be synthesized.
   */
  public readonly outdir: string;

  /**
   * The stack which will be synthesized. If not set, all stacks will be synthesized.
   */
  public readonly targetStackId: string | undefined;

  public readonly manifest: Manifest;

  /**
   * Whether to skip all validations during synthesis of the app
   */
  public readonly skipValidation: boolean;

  /**
   * Whether to skip backend validation during synthesis of the app
   */
  public readonly skipBackendValidation: boolean;

  /**
   * Defines an app
   * @param config configuration
   */
  constructor(config: AppConfig = {}) {
    super(undefined as any, "");
    Object.defineProperty(this, APP_SYMBOL, { value: true });

    this.outdir = process.env.CDKTF_OUTDIR ?? config.outdir ?? "cdktf.out";
    this.targetStackId = process.env.CDKTF_TARGET_STACK_ID;
    this.skipValidation = config.skipValidation || false;
    this.skipBackendValidation = config.skipBackendValidation || false;

    this.loadContext(config.context);

    const node = this.node;
    if (config.stackTraces === false) {
      node.setContext(DISABLE_STACK_TRACE_IN_METADATA, true);
    }

    node.setContext("cdktfVersion", version);

    if (!fs.existsSync(this.outdir)) {
      fs.mkdirSync(this.outdir);
    }
    this.manifest = new Manifest(version, this.outdir);
  }

  public static isApp(x: any): x is App {
    return x !== null && typeof x === "object" && APP_SYMBOL in x;
  }

  public static of(construct: IConstruct): App {
    return _lookup(construct);

    // eslint-disable-next-line jsdoc/require-jsdoc
    function _lookup(c: IConstruct): App {
      if (App.isApp(c)) {
        return c;
      }

      const node = c.node;

      if (!node.scope) {
        throw new Error(
          `No app could be identified for the construct at path '${construct.node.path}'`
        );
      }

      return _lookup(node.scope);
    }
  }

  /**
   * Synthesizes all resources to the output directory
   */
  public synth(): void {
    const session: ISynthesisSession = {
      outdir: this.outdir,
      skipValidation: this.skipValidation,
      manifest: this.manifest,
    };

    const stacks = this.node
      .findAll()
      .filter<TerraformStack>(TerraformStack.isStack);

    stacks.forEach((stack) => stack.prepareStack());
    stacks.forEach((stack) => stack.synthesizer.synthesize(session));

    if (!this.skipValidation) {
      const validations = this.node.validate();
      if (validations.length) {
        const errorList = validations.join("\n  ");
        throw new Error(
          `App level validation failed with the following errors:\n  ${errorList}`
        );
      }
    }

    this.manifest.writeToFile();
  }

  /**
   * Creates a reference from one stack to another, invoked on prepareStack since it creates extra resources
   */
  public crossStackReference(
    fromStack: TerraformStack,
    toStack: TerraformStack,
    identifier: string
  ): string {
    toStack.addDependency(fromStack);
    const outputId =
      fromStack.registerOutgoingCrossStackReference(
        identifier
      ).friendlyUniqueId;

    const remoteState = toStack.registerIncomingCrossStackReference(fromStack);

    return remoteState.getString(outputId);
  }

  private loadContext(defaults: { [key: string]: string } = {}) {
    const node = this.node;

    // prime with defaults passed through constructor
    for (const [k, v] of Object.entries(defaults)) {
      node.setContext(k, v);
    }

    // read from environment
    const contextJson = process.env[CONTEXT_ENV];
    const contextFromEnvironment = contextJson ? JSON.parse(contextJson) : {};

    for (const [k, v] of Object.entries(contextFromEnvironment)) {
      node.setContext(k, v);
    }
  }
}
