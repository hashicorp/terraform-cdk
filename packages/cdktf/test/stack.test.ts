// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import {
  TerraformResource,
  TerraformStack,
  App,
  Testing,
  TerraformOutput,
  LocalBackend,
} from "cdktf/lib";
import { TerraformModule } from "cdktf/lib/terraform-module";
import { TestProvider } from "./helper";
import { Construct } from "constructs";

test("stack synthesis merges all elements into a single output", () => {
  const app = Testing.stubVersion(
    Testing.enableFutureFlags(new App({ stackTraces: false }))
  );
  const stack = new TerraformStack(app, "MyStack");

  new TestProvider(stack, "test-provider", {
    accessKey: "foo",
    type: "aws",
  });

  new MyResource(stack, "Resource1", {
    terraformResourceType: "aws_bucket",
  });

  const overrideResource = new MyResource(stack, "Resource2", {
    terraformResourceType: "aws_topic",
  });
  overrideResource.addOverride("//", "this is a comment");
  overrideResource.addOverride("prop2", undefined);
  overrideResource.addOverride("prop3.name", "test");
  overrideResource.addOverride("provisioner", [
    {
      "local-exec": {
        command: "echo 'Hello World' >example.txt",
      },
    },
  ]);

  const eks = new MyModule(stack, "EksModule", {
    source: "terraform-aws-modules/eks/aws",
    version: "7.0.1",
  });

  new TerraformOutput(stack, "eks_version", {
    value: eks.version,
  });

  stack.addOverride("terraform.backend", {
    remote: {
      organization: "test",
      workspaces: {
        name: "test",
      },
    },
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("stack synthesis no flags", () => {
  const app = Testing.stubVersion(new App({ stackTraces: false }));
  const stack = new TerraformStack(app, "MyStack");

  new TestProvider(stack, "test-provider", {
    accessKey: "foo",
    type: "aws",
  });

  new MyResource(stack, "Resource1", {
    terraformResourceType: "aws_bucket",
  });

  const eks = new MyModule(stack, "EksModule", {
    source: "terraform-aws-modules/eks/aws",
    version: "7.0.1",
  });

  new TerraformOutput(stack, "eks_version", {
    value: eks.version,
  });

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test("stack validation returns error when provider is missing", () => {
  const app = Testing.stubVersion(new App({ stackTraces: false }));
  const stack = new TerraformStack(app, "MyStack");

  new MyResource(stack, "Resource1", {
    terraformResourceType: "aws_bucket",
    terraformGeneratorMetadata: {
      providerName: "test-provider",
    },
  });

  const errors = stack.node.validate();

  expect(errors).toEqual([
    `Found resources without a matching provider construct. Please make sure to add provider constructs [e.g. new RandomProvider(...)] to your stack 'MyStack' for the following providers: test-provider`,
  ]);
});

test("stack validation returns no error when provider is not set", () => {
  const app = Testing.stubVersion(new App({ stackTraces: false }));
  const stack = new TerraformStack(app, "MyStack");

  new MyResource(stack, "Resource1", {
    terraformResourceType: "aws_bucket",
  });

  const errors = stack.node.validate();
  expect(errors).toEqual([]);
});

test("getting Stack for construct which was added to root app returns friendly error", () => {
  const app = Testing.stubVersion(new App({ stackTraces: false }));
  new TerraformStack(app, "MyStack");

  expect(() => new LocalBackend(app, {})).toThrowErrorMatchingInlineSnapshot(
    `"No stack could be identified for the construct at path 'backend'. You seem to have passed your root App as scope to a construct. Pass a stack (inheriting from TerraformStack) as scope to your construct instead."`
  );
});

describe("output id map", () => {
  test("output id map is empty when no outputs are added", () => {
    const app = Testing.stubVersion(new App({ stackTraces: false }));
    const stack = new TerraformStack(app, "MyStack");

    expect(stack.toTerraform()["//"].outputs).toEqual({});
  });

  test("output id map is populated when outputs are added", () => {
    const app = Testing.stubVersion(new App({ stackTraces: false }));
    const stack = new TerraformStack(app, "MyStack");

    new TerraformOutput(stack, "output1", {
      value: "value1",
    });

    const resource = new MyResource(stack, "Resource1", {
      terraformResourceType: "aws_bucket",
    });

    new TerraformOutput(stack, "output2", {
      value: resource,
    });

    const tf = stack.toTerraform();
    const { outputs } = tf["//"];

    expect(Object.keys(outputs.MyStack)).toEqual(["output1", "output2"]);
    expect(Object.keys(tf.output)).toEqual(
      expect.arrayContaining(Object.values(outputs.MyStack))
    );
  });

  test("output id map is populated with nested custom constructs", () => {
    const app = Testing.stubVersion(new App({ stackTraces: false }));
    const stack = new TerraformStack(app, "MyStack");

    class MyCustomConstruct extends Construct {
      constructor(scope: Construct, id: string) {
        super(scope, id);

        const resource = new MyResource(this, "resource", {
          terraformResourceType: "aws_bucket",
        });

        new TerraformOutput(this, "output", {
          value: resource,
        });
        new TerraformOutput(this, "output1", {
          value: "value1",
        });
      }
    }

    new MyCustomConstruct(stack, "MyCustomConstruct");
    new MyCustomConstruct(stack, "MyCustomOtherConstruct");

    const tf = stack.toTerraform();
    const { outputs } = tf["//"];

    expect(outputs).toMatchInlineSnapshot(`
      {
        "MyStack": {
          "MyCustomConstruct": {
            "output": "MyCustomConstruct_output_AFA0492F",
            "output1": "MyCustomConstruct_output1_11AB14A2",
          },
          "MyCustomOtherConstruct": {
            "output": "MyCustomOtherConstruct_output_D26A811A",
            "output1": "MyCustomOtherConstruct_output1_9A3A64D2",
          },
        },
      }
    `);
  });
});

class MyModule extends TerraformModule {
  protected synthesizeAttributes() {
    return {
      cluster_name: "my_cluster_name",
    };
  }
}

class MyResource extends TerraformResource {
  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      foo: this.friendlyUniqueId,
      prop1: "bar1",
      prop2: 1234,
      prop3: {
        name: "should be overwritten in resource 2",
        value: 5678,
      },
    };
  }
}
