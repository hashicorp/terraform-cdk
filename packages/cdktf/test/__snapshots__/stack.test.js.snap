// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stack synthesis merges all elements into a single output 1`] = `
"{
  "module": {
    "EksModule": {
      "cluster_name": "my_cluster_name",
      "source": "terraform-aws-modules/eks/aws",
      "version": "7.0.1"
    }
  },
  "output": {
    "eks_version": {
      "value": "7.0.1"
    }
  },
  "provider": {
    "aws": [
      {
        "access_key": "foo"
      }
    ]
  },
  "resource": {
    "aws_bucket": {
      "Resource1": {
        "foo": "Resource1",
        "prop1": "bar1",
        "prop2": 1234,
        "prop3": {
          "name": "should be overwritten in resource 2",
          "value": 5678
        }
      }
    },
    "aws_topic": {
      "Resource2": {
        "foo": "Resource2",
        "prop1": "bar1",
        "prop3": {
          "name": "test",
          "value": 5678
        },
        "provisioner": [
          {
            "local-exec": {
              "command": "echo 'Hello World' >example.txt"
            }
          }
        ]
      }
    }
  },
  "terraform": {
    "backend": {
      "remote": {
        "organization": "test",
        "workspaces": {
          "name": "test"
        }
      }
    },
    "required_providers": {
      "aws": {
        "version": "~> 2.0"
      }
    }
  }
}"
`;

exports[`stack synthesis no flags 1`] = `
"{
  "module": {
    "EksModule": {
      "cluster_name": "my_cluster_name",
      "source": "terraform-aws-modules/eks/aws",
      "version": "7.0.1"
    }
  },
  "output": {
    "eks_version": {
      "value": "7.0.1"
    }
  },
  "provider": {
    "aws": [
      {
        "access_key": "foo"
      }
    ]
  },
  "resource": {
    "aws_bucket": {
      "Resource1": {
        "foo": "Resource1",
        "prop1": "bar1",
        "prop2": 1234,
        "prop3": {
          "name": "should be overwritten in resource 2",
          "value": 5678
        }
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "version": "~> 2.0"
      }
    }
  }
}"
`;
